### PR TITLE
fixes issue #156; #158

### LINKED ISSUE #156: make streams.publish return its arguments when it is successful
void procedures are hard to chain together with other stuff in cypher.
Thanks to @moxious

### LINKED ISSUE #158: streams.produce doesn't publish in case of null argument
CALL streams.publish(‘test’, null) does not publish { payload: null }.
Provide an exception in order to highlight to the user the behaviour.

### PR DESCRIPTION
**Pull request by moxious**
Fixes #156 
Fixes #158 

CALL streams.publish(a, b, c) returns its arguments when it is successful, allowing verification that the message was published, and chaining of arguments in cypher.

### CODE REVIEW

**Comment by moxious** [comment]
Achievement unlocked -- my first neo4j-streams PR and my first ever written line of Kotlin.  :)

File: producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt [hunk 1]
```diff
@@ -15,15 +27,15 @@ class StreamsProcedures {
     @Procedure(mode = Mode.SCHEMA, name = "streams.publish")
     @Description("streams.publish(topic, config) - Allows custom streaming from Neo4j to the configured stream environment")
     fun publish(@Name("topic") topic: String?, @Name("payload") payload: Any?,
-                @Name(value = "config", defaultValue = "{}") config: Map<String, Any>?) {
+                @Name(value = "config", defaultValue = "{}") config: Map<String, Any>?): Stream<PublishResult> {
         checkEnabled()
         if (topic.isNullOrEmpty()) {
             log?.info("Topic empty, no message sent")
```
Comment (jexp) [inline producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: should we rather return this to the client

File: producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt [hunk 2]
```diff
@@ -38,6 +50,8 @@ class StreamsProcedures {
                 .withTopic(topic)
                 .build()
         StreamsProcedures.eventRouter.sendEvents(topic, listOf(streamsEvent))
```
Comment (jexp) [inline producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: this is asynchronous, afaik 

so we're not sure if the event was actually sent, it's more like fire&forget.

File: producer/src/test/kotlin/streams/integrations/KafkaEventRouterIT.kt [hunk 3]
```diff
@@ -154,7 +156,15 @@ class KafkaEventRouterIT {
...
         val message = "Hello World"
-        db.execute("CALL streams.publish('neo4j', '$message')").close()
+
+        val result = db.execute("CALL streams.publish('neo4j', '$message')")
+        assertTrue { result.hasNext() }
+        val resultMap = result.next()
+
+        assertTrue { "$message".equals(resultMap.get("payload")) }
+        assertTrue { "neo4j".equals(resultMap.get("topic")) }
+
```
Comment (jexp) [inline producer/src/test/kotlin/streams/integrations/KafkaEventRouterIT.kt]: assertFalse result.hasNext

File: producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt (hunk 2 above)
Comment (jexp) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: @conker84 wdyt

Comment (conker84) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: Yes it's an async call. 
It would be cool to add a new method `sendSync` into StreamsEventRouter and implement it in the `KafkaEventRouter` in this way:

```kotlin
    override fun startSync(producerRecord: ProducerRecord<String, ByteArray>): RecordMetadata {
        producer.send(producerRecord).get()
    }
``` 

So we can provide as output the `RecordMetadata`

Comment (moxious) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: ah you're right this is even better.  I'll look into it.

File: producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt (hunk 1 above)
Comment (moxious) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: I don't understand this comment.  Do you mean we should not return config, or do you mean we should find a way to return the log message if topic is empty?

Comment (jexp) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: I meant that no one will see the log. And if we're already returning stuff we should probably return this info (that it was actually not sent instead of an empty stream).

Comment (moxious) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: You are right -- this is existing code and I would like to keep it to a different PR some other time though for several reasons -- 

When you are sending multiple, it's ambiguous whether you mean "best effort" deliver as many as you can, or if you mean fail if one error happens.  Also if one error happens on number 7, you can't recall 1 - 6.   Error handling here isn't in the scope of the issues I'm trying to tackle, but it is something we should maybe improve.

To overcome the limitations of the semantics of this call, I now have implemented a synchronous send method, which won't succeed unless all are sent, and you get back RecordMetadata.  This is what streams.publish will now use, which has the nice side effect of permitting us to return record metadata from kafka.

File: producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt (hunk 2 above)
Comment (moxious) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: Pushed updates.  @conker84 had implemented "fire and forget".  I changed the return types of several functions to return Futures.  The fire and forget behavior of this method has been retained, **but** I have implemented a new sendSync method according to @conker84 's suggestion.  I changed streams.publish to use this, so now it can return record metadata to the client.

"Fire and forget" is still possible with the API, and at a later point we might want to make this a config option.  Default behavior for streams.publish (in my opinion) should be to wait until it can confirm it did what you asked.  This matches the behavior of streams.consume which polls / blocks.

File: producer/src/main/kotlin/streams/kafka/KafkaEventRouter.kt [hunk 4]
```diff
@@ -42,43 +43,44 @@ class KafkaEventRouter: StreamsEventRouter {
...
 
-    override fun sendEvents(topic: String, transactionEvents: List<out StreamsEvent>) {
-        try {
+    override fun sendEvents(topic: String, transactionEvents: List<out StreamsEvent>): List<out Future<RecordMetadata>> {
+        val sendFutures: MutableList<Future<RecordMetadata>> = mutableListOf()
+
+         try {
             producer.beginTransaction()
             transactionEvents.forEach {
+                // Use a mutable list and add each one, because we might send some and get an exception
```
Comment (moxious) [inline producer/src/main/kotlin/streams/kafka/KafkaEventRouter.kt]: TODO -- how should sendEvents behave if you fire & forget 6, but the third fails?

This change keeps present behavior intact, which is that it tries to do all 6.  A failure may stop some intermediate result, but it still does "best effort" on all of them.

File: producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt [hunk 5]
```diff
@@ -37,7 +62,9 @@ class StreamsProcedures {
                         .firstOrNull())
                 .withTopic(topic)
                 .build()
-        StreamsProcedures.eventRouter.sendEvents(topic, listOf(streamsEvent))
+        val result = StreamsProcedures.eventRouter.sendEventsSync(topic, listOf(streamsEvent))
```
Comment (moxious) [inline producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: Now using synchronous behavior by default which permits sending RecordMetadata.

File: producer/src/test/kotlin/streams/integrations/KafkaEventRouterIT.kt [hunk 6]
```diff
@@ -166,6 +182,17 @@ class KafkaEventRouterIT {
         consumer.close()
     }
 
+    @Test
+    fun testCantPublishNull() {
```
Comment (moxious) [inline producer/src/test/kotlin/streams/integrations/KafkaEventRouterIT.kt]: @conker84  in a comment said that we should document this behavior, so now we throw an Exception and verify that is the case.

I wonder if we should allow publishing null though....I hear some people use this in kafka as a "tombstone record", but not sure if that applies here.  I'm new to this repo so I'm sticking with current behavior until we have a different requirement.

File: producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt (hunk 2 above)
Comment (conker84) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: I was wondering if instead of returning the `RecordMetadata`, it would be better to return a meta-structure that should be general as possible. This because we designed the Stream module in order to provide different "streaming connectors" (kinesis, rabbitMQ, ...)
@jexp @moxious wdyt?

Comment (moxious) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: If you like, a solution here is I could return record metadata as a map.  In this way I can include all the kafka-specific metadata, but the return type of the procedure can always be the same.  If it's a different streaming tech, you still get recordMetadata as a result, but the map has different keys.

Comment (conker84) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: Cool!

Comment (moxious) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: @jexp @conker84 I'm going down a deep rabbit hole here and might need some help.

So I created a wrapper class PublishedRecordMetadata and then implemented that in a specific case for kafka to make this generic.  However the Kafka publish functions all return Future<RecordMetadata>.   Doing research, these Futures are not pipe-lineable.  (And so doing the trivial bit of turning Future<RecordMetadata> into Future<KafkaPublishedRecordMetadata> is turning out to be stupid complex)

 In the Java world they're telling me that I need a CompletableFuture because Future is the old API.  In the Kotlin world they're telling me that I should use something else entirely.

So basically this simple suggestion has turned into something much more complicated for me with Kotlin/java async programming.

What is your opinion on how this should be done?  Forgetting for a moment the RecordMetadata class, what async implementation do you want this library to return e.g. return type here:

```
fun send(producerRecord: ProducerRecord<String, ByteArray>): Future<PublishedRecordMetadata>
```

I don't want to prep the PR again and find out that it needs to be totally different . :)

Comment (conker84) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: Maybe we can consider to keep it simple and change the _old_ `sendEvent` method in orderd to take a `callback` as third argument (so it can still be used with the fire-and-forget style); at this point the new `sendEventSync` should simply returns a `List<PublishedRecordMetadata>`

wdyt?

Comment (moxious) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: Not sure I understand how that would work.   Suppose I took a callback, I'm still going to get a `Future<RecordMetadata>` from the Kafka API.  If I write some code to `.get()` the value from that future, it's going to block in the async method -- or else if it doesn't block, then passing a callback is really the same as just calling `.get()` on the future yourself.

Isn't it also sort of the case that adding callbacks to APIs tends to make them messier and encourage "pyramid" style programming, I think Futures and promises were intended to get around this.   It does seem like a nasty combination of Java/Kotlin issues here though.

File: producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt [hunk 7]
```diff
@@ -15,15 +40,15 @@ class StreamsProcedures {
...
         checkEnabled()
         if (topic.isNullOrEmpty()) {
             log?.info("Topic empty, no message sent")
-            return
+            return Stream.empty();
         }
         if (payload == null) {
             log?.info("Payload empty, no message sent")
-            return
+            throw RuntimeException("Payload may not be null")
```
Comment (jexp) [inline producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: don't do that. imagine you have many messages then one null will result in a half-send batch

File: producer/src/test/kotlin/streams/mocks/MockStreamsEventRouter.kt [hunk 8]
```diff
@@ -1,15 +1,44 @@
...
+            val future = CompletableFuture<RecordMetadata>()
+            future.complete(fakeRecordMetadata(topic))
+            result.add(future)
+        }
+
+        return result
+    }
+
+    override fun sendEventsSync(topic: String, transactionEvents: List<out StreamsEvent>): List<out RecordMetadata> {
+        var result = mutableListOf<RecordMetadata>()
```
Comment (jexp) [inline producer/src/test/kotlin/streams/mocks/MockStreamsEventRouter.kt]: transacations.map

File: producer/src/test/kotlin/streams/mocks/MockStreamsEventRouter.kt [hunk 9]
```diff
@@ -1,15 +1,44 @@
...
+    fun fakeRecordMetadata(topic: String): RecordMetadata {
+        return RecordMetadata(
+                TopicPartition(topic, 0),
+                0, 1, 2, 3, 4, 5)
+    }
+
+    override fun sendEvents(topic: String, streamsTransactionEvents: List<out StreamsEvent>): List<out Future<RecordMetadata>> {
         events.addAll(streamsTransactionEvents as List<StreamsTransactionEvent>)
+
+        val result = mutableListOf<Future<RecordMetadata>>()
```
Comment (jexp) [inline producer/src/test/kotlin/streams/mocks/MockStreamsEventRouter.kt]: streamsTranscations.map

File: producer/src/test/kotlin/streams/mocks/MockStreamsEventRouter.kt [hunk 10]
```diff
@@ -1,15 +1,44 @@
...
 import streams.events.StreamsEvent
 import streams.events.StreamsTransactionEvent
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.Future
 
 class MockStreamsEventRouter(logService: LogService? = null, config: Config? = null): StreamsEventRouter(logService, config) {
 
-    override fun sendEvents(topic: String, streamsTransactionEvents: List<out StreamsEvent>) {
+    fun fakeRecordMetadata(topic: String): RecordMetadata {
+        return RecordMetadata(
```
Comment (jexp) [inline producer/src/test/kotlin/streams/mocks/MockStreamsEventRouter.kt]: no need for return type + `return` just `= RecordMetadata(..`

File: producer/src/main/kotlin/streams/kafka/KafkaEventRouter.kt [hunk 11]
```diff
@@ -42,43 +43,44 @@ class KafkaEventRouter: StreamsEventRouter {
...
         }
         val producerRecord = ProducerRecord(topic, partition, System.currentTimeMillis(), "${event.meta.txId + event.meta.txEventId}-${event.meta.txEventId}",
                 JSONUtils.writeValueAsBytes(event))
-        send(producerRecord)
+        return send(producerRecord)
     }
 
-    override fun sendEvents(topic: String, transactionEvents: List<out StreamsEvent>) {
-        try {
+    override fun sendEvents(topic: String, transactionEvents: List<out StreamsEvent>): List<out Future<RecordMetadata>> {
```
Comment (jexp) [inline producer/src/main/kotlin/streams/kafka/KafkaEventRouter.kt]: Instead using futures I would want to keep this `async` and then use `use await sendEvents()` in the sync method.
@conker84 will look into this.

File: producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt (hunk 2 above)
Comment (conker84) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: I think we should convert the `Future` in a `Deferred` by using (the Kotlin extension function [feature](https://kotlinlang.org/docs/reference/extensions.html)) in this way:
```kotlin
fun <T> Future<T>.asDeferred(dispatcher: CoroutineDispatcher = Dispatchers.Default): Deferred<T> {
    return GlobalScope.async(dispatcher) {
        get()
    }
}
```

An use it in that way:
```kotlin
private fun send(producerRecord: ProducerRecord<String, ByteArray>): Deferred<RecordMetadata> {
  return producer.send(producerRecord) { meta: RecordMetadata?, error: Exception? ->
    if (log.isDebugEnabled) {
      log.debug("Sent record in partition ${meta?.partition()} offset ${meta?.offset()} data ${meta?.topic()} key size ${meta?.serializedKeySize()}", error)
    }
  }.asDeferred()
}
```

Comment (conker84) [reply producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt]: I had a chat on the Kotlin Slack and basically, we have two options:
- change the implementation of `asDeferred` in this way:
```
fun <T> Future<T>.asDeferred(): Deferred<T> {
    return GlobalScope.async(Dispatchers.IO) {
        get()
    }
}
```
- implement a solution as described [here](https://stackoverflow.com/questions/50793362/how-to-use-suspendcoroutine-to-turn-java-7-future-into-kotlin-suspending-functio#answer-52901712)

**Comment by moxious** [comment]
Too many issues, aged out.
